rust/azure_iot_operations_services/src/azure_device_registry/client.rs

```diff
@@ -1,2 +1,1608 @@
+use azure_iot_operations_protocol::application::ApplicationContext;
+use derive_builder::Builder;
+use tokio::sync::Notify;
+
+use crate::azure_device_registry::device_name_gen::adr_base_service::client as adr_name_gen;
+use crate::azure_device_registry::{
+    Asset, AssetStatus, AssetUpdateObservation, Device, DeviceStatus, DeviceUpdateObservation,
+    Error, ErrorKind,
+    device_name_gen::{
+        common_types::options::CommandInvokerOptionsBuilder,
+        common_types::options::TelemetryReceiverOptionsBuilder,
+    },
+};
+use crate::common::dispatcher::{DispatchError, Dispatcher};
+
+const DEVICE_NAME_TOPIC_TOKEN: &str = "deviceName";
+const DEVICE_NAME_RECEIVED_TOPIC_TOKEN: &str = "ex:deviceName";
+const INBOUND_ENDPOINT_NAME_TOPIC_TOKEN: &str = "inboundEndpointName";
+const INBOUND_ENDPOINT_NAME_RECEIVED_TOPIC_TOKEN: &str = "ex:inboundEndpointName";
+
+/// Options for the Azure Device Registry client.
+#[derive(Builder, Clone, Default)]
+#[builder(setter(into))]
+pub struct ClientOptions {
+    /// If true, update notifications are auto-acknowledged
+    #[builder(default = "true")]
+    notification_auto_ack: bool,
+}
+
+/// Azure Device Registry client implementation.
+#[derive(Clone)]
+pub struct Client<C>
+where
+    C: ManagedClient + Clone + Send + Sync + 'static,
+    C::PubReceiver: Send + Sync,
+{
+    // general
+    shutdown_notifier: Arc<Notify>,
+    // device
+    get_device_command_invoker: Arc<adr_name_gen::GetDeviceCommandInvoker<C>>,
+    update_device_status_command_invoker: Arc<adr_name_gen::UpdateDeviceStatusCommandInvoker<C>>,
+    notify_on_device_update_command_invoker:
+        Arc<adr_name_gen::SetNotificationPreferenceForDeviceUpdatesCommandInvoker<C>>,
+    device_update_notification_dispatcher: Arc<Dispatcher<(Device, Option<AckToken>)>>,
+    // asset
+    get_asset_command_invoker: Arc<adr_name_gen::GetAssetCommandInvoker<C>>,
+    update_asset_status_command_invoker: Arc<adr_name_gen::UpdateAssetStatusCommandInvoker<C>>,
+    notify_on_asset_update_command_invoker:
+        Arc<adr_name_gen::SetNotificationPreferenceForAssetUpdatesCommandInvoker<C>>,
+    asset_update_notification_dispatcher: Arc<Dispatcher<(Asset, Option<AckToken>)>>,
+}
+
+impl<C> Client<C>
+where
+    C: ManagedClient + Clone + Send + Sync + 'static,
+    C::PubReceiver: Send + Sync,
+{
+    // ~~~~~~~~~~~~~~~~~ General APIs ~~~~~~~~~~~~~~~~~~~~~
+    /// Create a new Azure Device Registry Client.
+    ///
+    /// # Errors
+    /// [`struct@Error`] of kind [`InvalidClientId`](ErrorKind::InvalidClientId)
+    /// if the Client Id of the [`ManagedClient`] isn't valid as a topic token.
+    #[allow(clippy::needless_pass_by_value)]
+    pub fn new(
+        application_context: ApplicationContext,
+        client: C,
+        options: ClientOptions,
+    ) -> Result<Self, Error> {
+        if !Self::is_valid_replacement(client.client_id()) {
+            return Err(ErrorKind::InvalidClientId(client.client_id().to_string()).into());
```

Comment:
This error looks like it's expecting an error message for the String, not just the field with the issue. Modify either here or in the error definition to include an error message

Reply:
Also not sure if the existing from impls for this errorKind still make sense now

Reply:
The exisitng impls are needed as by default we are getting `Result<CommandInvokerOptions, CommandInvokerOptionsBuilderError>` on build. To avoid that we have to either add a panic or an else block to handle that error. I have kept the panic for now.